pkg/beacon/relay/entry/entry.go

```diff
@@ -33,26 +39,195 @@ func SignAndSubmit(
 	signer *dkg.ThresholdSigner,
 	startBlockHeight uint64,
 ) error {
+	ctx, cancelCtx := context.WithCancel(context.Background())
+	defer cancelCtx()
+
+	relayEntrySubmittedChannel := make(chan uint64)
+	subscription, err := relayChain.OnRelayEntrySubmitted(
+		func(event *event.EntrySubmitted) {
+			relayEntrySubmittedChannel <- event.BlockNumber
+		},
+	)
+	if err != nil {
+		return err
+	}
+	defer subscription.Unsubscribe()
+
+	chainConfig, err := relayChain.GetConfig()
+	if err != nil {
+		return err
+	}
+
+	relayEntryTimeoutChannel, err := blockCounter.BlockHeightWaiter(
+		startBlockHeight + chainConfig.RelayEntryTimeout,
+	)
+	if err != nil {
+		return err
+	}
+
 	previousEntry := new(bn256.G1)
-	_, err := previousEntry.Unmarshal(previousEntryBytes)
+	_, err = previousEntry.Unmarshal(previousEntryBytes)
 	if err != nil {
 		return err
 	}
 
-	initialState := &signatureShareState{
-		signingStateBase: signingStateBase{
-			channel:         channel,
-			relayChain:      relayChain,
-			blockCounter:    blockCounter,
-			signer:          signer,
-			previousEntry:   previousEntry,
-			honestThreshold: honestThreshold,
-		},
-		signingStartBlockHeight: startBlockHeight,
+	selfShare := signer.CalculateSignatureShare(previousEntry)
+
+	go broadcastShare(ctx, signer.MemberID(), selfShare, channel)
```

Comment:
I see know how much that broadcast channel + ctx refactor was a good idea :heart:

Reply:
Indeed :)